### ISSUE TITLE
Error "<type> is not in the type environment at a reify" when using mkParseJSON/mkToEncoding on a type in the same module.

### ISSUE THREAD

**Issue by thestr4ng3r** [bos/aeson#797]
In the following example (using aeson 1.5.2.0 and ghc 8.10.1):
```haskell
{-# LANGUAGE TemplateHaskell #-}
module Main where

import Data.Aeson.Types
import Data.Aeson.TH

data SomeType = Something

parseSomeType :: Value -> Parser SomeType
parseSomeType = $(mkParseJSON defaultOptions ''SomeType)
```

I get this error:
```
Main.hs:10:19: error:
    • ‘SomeType’ is not in the type environment at a reify
    • In the untyped splice: $(mkParseJSON defaultOptions ''SomeType)
   |
10 | parseSomeType = $(mkParseJSON defaultOptions ''SomeType)
   |
```

`deriveJSON` et al. work fine here. Interestingly, `mkParseJSON` does work if `SomeType` is defined in some other module and only imported, which is what is covered by the tests in https://github.com/bos/aeson/blob/master/tests/Encoders.hs

I'm still quite new to Template Haskell and have no idea what the error means. Maybe I'm doing something wrong too but it seems like `mkParseJSON` should be quite straightforward to use.

**Comment by bergmark** [comment]
It's been a while since i did TH now, but I think this is just a limitation of TH itself. If I'm right then there's not much we can do in aeson.

**Comment by thestr4ng3r** [comment]
I think the limitation of TH wrt. modules would be that you can't call a function in splices unless it's imported. What's interesting here is that `deriveJSON` does work in the same context, with the main difference being that it generates the class instance which would theoretically include everything that `mkParseJSON` does.
But I should look into the code myself more.

**Comment by thestr4ng3r** [comment]
Ok you were right, the issue is discussed in detail here: https://gitlab.haskell.org/ghc/ghc/-/issues/9813 and documented here: https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#th-syntax under `Top-level declaration splices break up a source file into declaration groups...`

tl;dr
Expression slices don't break up declaration groups and thus a direct `$(mkParseJSON defaultOptions ''SomeType)` won't see `SomeType`
Top-level declaration slices like `$(deriveJSON defaultOptions ''SomeType` do break up declaration groups so `SomeType` will be visible before it.
The fix is to put some top-level declaration splice before the expression splice, in the simplest case just `$(return [])`:
```haskell
{-# LANGUAGE TemplateHaskell #-}
module Main where

import Data.Aeson.Types
import Data.Aeson.TH

data SomeType = Something

$(return [])

parseSomeType :: Value -> Parser SomeType
parseSomeType = $(mkParseJSON defaultOptions ''SomeType)
```

I think this can be quite useful info to other users in the future.
